File: ios/TimerData.swift
import Foundation

// ─── Timer Status ─────────────────────────────────────────────────────────────

@objc enum TimerStatus: Int {
  case idle
  case running
  case paused
  case completed
  case stopped

  var stringValue: String {
    switch self {
    case .idle:      return "idle"
    case .running:   return "running"
    case .paused:    return "paused"
    case .completed: return "completed"
    case .stopped:   return "stopped"
    }
  }

  static func from(_ string: String) -> TimerStatus {
    switch string {
    case "running":   return .running
    case "paused":    return .paused
    case "completed": return .completed
    case "stopped":   return .stopped
    default:          return .idle
    }
  }
}

// ─── Timer Model ──────────────────────────────────────────────────────────────

struct TimerData: Codable {
  var id: String
  var status: String           // uses string for Codable simplicity
  var duration: TimeInterval   // ms
  var elapsed: TimeInterval    // ms
  var remaining: TimeInterval  // ms
  var tickInterval: TimeInterval
  var startedAt: TimeInterval  // epoch ms (0 = never)
  var completedAt: TimeInterval
  var pausedAt: TimeInterval
  var persist: Bool
  var iosTaskIdentifier: String

  var timerStatus: TimerStatus { TimerStatus.from(status) }

  init(
    id: String,
    status: TimerStatus = .idle,
    duration: TimeInterval,
    elapsed: TimeInterval = 0,
    remaining: TimeInterval? = nil,
    tickInterval: TimeInterval = 1000,
    startedAt: TimeInterval = 0,
    completedAt: TimeInterval = 0,
    pausedAt: TimeInterval = 0,
    persist: Bool = true,
    iosTaskIdentifier: String = ""
  ) {
    self.id = id
    self.status = status.stringValue
    self.duration = duration
    self.elapsed = elapsed
    self.remaining = remaining ?? duration
    self.tickInterval = tickInterval
    self.startedAt = startedAt
    self.completedAt = completedAt
    self.pausedAt = pausedAt
    self.persist = persist
    self.iosTaskIdentifier = iosTaskIdentifier.isEmpty
      ? "com.rick427.backgroundtimer.\(id)"
      : iosTaskIdentifier
  }

  /// Convert to the dictionary format expected by React Native
  func toRNMap() -> [String: Any] {
    return [
      "id": id,
      "status": status,
      "duration": duration,
      "elapsed": elapsed,
      "remaining": remaining,
      "startedAt": startedAt,
      "completedAt": completedAt,
      "pausedAt": pausedAt,
    ]
  }
}

File: android/src/main/java/com/rick427/backgroundtimer/TimerData.kt
package com.rick427.backgroundtimer

import org.json.JSONObject

/**
 * Immutable snapshot of a single timer's state.
 * Serialized to/from JSON for SharedPreferences persistence.
 */
data class TimerData(
  val id: String,
  val status: TimerStatus,
  val duration: Long,          // ms
  val elapsed: Long,           // ms
  val remaining: Long,         // ms
  val tickInterval: Long,      // ms (0 = no ticks)
  val startedAt: Long,         // epoch ms (0 = never started)
  val completedAt: Long,       // epoch ms (0 = not completed)
  val pausedAt: Long,          // epoch ms (0 = not paused)
  val persist: Boolean,
  val androidConfig: AndroidNotificationConfig,
  val iosTaskIdentifier: String,
) {
  fun toJson(): JSONObject = JSONObject().apply {
    put("id", id)
    put("status", status.value)
    put("duration", duration)
    put("elapsed", elapsed)
    put("remaining", remaining)
    put("tickInterval", tickInterval)
    put("startedAt", startedAt)
    put("completedAt", completedAt)
    put("pausedAt", pausedAt)
    put("persist", persist)
    put("androidConfig", androidConfig.toJson())
    put("iosTaskIdentifier", iosTaskIdentifier)
  }

  companion object {
    fun fromJson(json: JSONObject): TimerData = TimerData(
      id = json.getString("id"),
      status = TimerStatus.fromValue(json.getString("status")),
      duration = json.getLong("duration"),
      elapsed = json.getLong("elapsed"),
      remaining = json.getLong("remaining"),
      tickInterval = json.getLong("tickInterval"),
      startedAt = json.getLong("startedAt"),
      completedAt = json.getLong("completedAt"),
      pausedAt = json.getLong("pausedAt"),
      persist = json.getBoolean("persist"),
      androidConfig = AndroidNotificationConfig.fromJson(
        json.getJSONObject("androidConfig")
      ),
      iosTaskIdentifier = json.getString("iosTaskIdentifier"),
    )
  }
}

enum class TimerStatus(val value: String) {
  IDLE("idle"),
  RUNNING("running"),
  PAUSED("paused"),
  COMPLETED("completed"),
  STOPPED("stopped");

  companion object {
    fun fromValue(value: String): TimerStatus =
      entries.firstOrNull { it.value == value } ?: IDLE
  }
}

data class AndroidNotificationConfig(
  val title: String,
  val text: String,
  val icon: String,
  val color: String,
  val importance: String,
  val channelId: String,
  val channelName: String,
) {
  fun toJson(): JSONObject = JSONObject().apply {
    put("title", title)
    put("text", text)
    put("icon", icon)
    put("color", color)
    put("importance", importance)
    put("channelId", channelId)
    put("channelName", channelName)
  }

  companion object {
    val DEFAULT = AndroidNotificationConfig(
      title = "Timer Running",
      text = "A background timer is active",
      icon = "ic_notification",
      color = "#000000",
      importance = "low",
      channelId = "background_timer_channel",
      channelName = "Background Timer",
    )

    fun fromJson(json: JSONObject): AndroidNotificationConfig = AndroidNotificationConfig(
      title = json.optString("title", DEFAULT.title),
      text = json.optString("text", DEFAULT.text),
      icon = json.optString("icon", DEFAULT.icon),
      color = json.optString("color", DEFAULT.color),
      importance = json.optString("importance", DEFAULT.importance),
      channelId = json.optString("channelId", DEFAULT.channelId),
      channelName = json.optString("channelName", DEFAULT.channelName),
    )

    fun fromJsonString(jsonString: String): AndroidNotificationConfig =
      if (jsonString.isBlank()) DEFAULT
      else runCatching { fromJson(JSONObject(jsonString)) }.getOrDefault(DEFAULT)
  }
}
